### PR TITLE
Improve log clarity about missing pod keys for provider

### LINKED ISSUE #599: Unable to find pod "@"
## Bug Report

When installing Maesh and applying a configuration on a fresh cluster I often see weid errors saying that we couldn't find a pod "@". This name means the pod has no name and no namespace (in the Topology). 
I wondering why we sometimes get these errors, and how we end up with a Pod with no name and namespace. 

### What did you do?

- Install Maesh using Helm with ACL enabled

```
helm install maesh helm/chart/maesh \
  --namespace=maesh \
  --set image.pullPolicy=IfNotPresent \
  --set image.name=maesh-dev \
  --set image.tag=latest \
  --set limits.tcp=20 \
  --set limits.http=20 \
  --set smi.deploy=true \
  --set acl=true 
  --set controller.logLevel=debug 
  --set mesh.logLevel=debug
```

- Apply the manifests attached to this issue (TrafficTarget with a client Pod and a HTTP Service)

```
kubectl apply -f ./manifests/
```

### What did you expect to see?

I expected to see no errors in the logs

### What did you see instead?

I got this following log, complaining about a Pod `@`:

```
time="2020-05-25T07:59:03Z" level=error msg="Unable to find Pod \"@\""
```

### Output of controller log: (_What version of Maesh are you using?_)

```
time="2020-05-25T07:56:35Z" level=info msg="Configuration loaded from flags."
time="2020-05-25T07:56:35Z" level=debug msg="Starting maesh prepare..."
time="2020-05-25T07:56:35Z" level=debug msg="Using masterURL: \"\""
time="2020-05-25T07:56:35Z" level=debug msg="Using kubeconfig: \"\""
time="2020-05-25T07:56:35Z" level=info msg="Creating in-cluster client"
time="2020-05-25T07:56:35Z" level=debug msg="Building Kubernetes Client..."
time="2020-05-25T07:56:35Z" level=debug msg="Building SMI Access Client..."
time="2020-05-25T07:56:35Z" level=debug msg="Building SMI Specs Client..."
time="2020-05-25T07:56:35Z" level=debug msg="Building SMI Split Client..."
time="2020-05-25T07:56:35Z" level=info msg="Checking DNS provider"
time="2020-05-25T07:56:35Z" level=info msg="Checking CoreDNS"
time="2020-05-25T07:56:35Z" level=info msg="CoreDNS match"
time="2020-05-25T07:56:35Z" level=debug msg="ACL mode enabled: true"
time="2020-05-25T07:56:35Z" level=debug msg="Initializing API"
time="2020-05-25T07:56:35Z" level=debug msg="Initializing mesh controller"
time="2020-05-25T07:56:35Z" level=debug msg="Starting API"
time="2020-05-25T07:56:35Z" level=debug msg="Starting Informers"
time="2020-05-25T07:56:35Z" level=debug msg="API readiness enabled"
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:01Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:03Z" level=error msg="Unable to find Pod \"@\""
time="2020-05-25T07:59:03Z" level=error msg="Unable to find Pod \"@\""
```

### What is your environment & configuration (arguments, provider, platform, ...)?

Version: master
Platform: k3s running on k3d
arguments: 
```
helm install maesh helm/chart/maesh \
  --namespace=maesh \
  --set image.pullPolicy=IfNotPresent \
  --set image.name=maesh-dev \
  --set image.tag=latest \
  --set limits.tcp=20 \
  --set limits.http=20 \
  --set smi.deploy=true \
  --set acl=true \
  --set controller.logLevel=debug \
  --set mesh.logLevel=debug
```

### If applicable, please paste the yaml objects required to reproduce your issue
[manifests.tar.gz](https://github.com/containous/maesh/files/4676027/manifests.tar.gz)

### PR DESCRIPTION
**Pull request by walbertus**
## What does this PR do?

This PR: Improve logging clarity for provider.go. This will give a better understanding to debug issue #599

### How to test it

 - Reproduce the linked issue
or
 - `make test`

### CODE REVIEW

File: pkg/provider/provider.go [hunk 1]
```diff
@@ -611,10 +611,15 @@ func (p Provider) buildUDPEntrypoint(svc *topology.Service, port int32) (string,
 func (p *Provider) buildHTTPServiceFromService(t *topology.Topology, svc *topology.Service, scheme string, port int32) *dynamic.Service {
 	var servers []dynamic.Server
 
+	containMissingPod := false
+
 	for _, podKey := range svc.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:617]: In my opinion, it's better to add more information to the current log statement than to add an extra logic. Something like the following:

```
p.logger.Errorf("Unable to find Pod %q for HTTP service %s@%s", podKey, topology.Key{Name: svc.Name, Namespace: svc.Namespace})
```

File: pkg/provider/provider.go [hunk 2]
```diff
@@ -636,10 +645,15 @@ func (p *Provider) buildHTTPServiceFromService(t *topology.Topology, svc *topolo
 func (p *Provider) buildHTTPServiceFromTrafficTarget(t *topology.Topology, tt *topology.ServiceTrafficTarget, scheme string, port int32) *dynamic.Service {
 	servers := make([]dynamic.Server, len(tt.Destination.Pods))
 
+	containMissingPod := false
+
 	for i, podKey := range tt.Destination.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:642]: See my previous comment.

File: pkg/provider/provider.go [hunk 3]
```diff
@@ -659,10 +677,15 @@ func (p *Provider) buildHTTPServiceFromTrafficTarget(t *topology.Topology, tt *t
 func (p *Provider) buildTCPServiceFromService(t *topology.Topology, svc *topology.Service, port int32) *dynamic.TCPService {
 	var servers []dynamic.TCPServer
 
+	containMissingPod := false
+
 	for _, podKey := range svc.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:665]: See my previous comment.

File: pkg/provider/provider.go [hunk 4]
```diff
@@ -683,10 +710,15 @@ func (p *Provider) buildTCPServiceFromService(t *topology.Topology, svc *topolog
 func (p *Provider) buildUDPServiceFromService(t *topology.Topology, svc *topology.Service, port int32) *dynamic.UDPService {
 	var servers []dynamic.UDPServer
 
+	containMissingPod := false
+
 	for _, podKey := range svc.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:689]: See my previous comment.

File: pkg/provider/provider.go [hunk 5]
```diff
@@ -707,16 +743,25 @@ func (p *Provider) buildUDPServiceFromService(t *topology.Topology, svc *topolog
 func (p *Provider) buildTCPServiceFromTrafficTarget(t *topology.Topology, tt *topology.ServiceTrafficTarget, port int32) *dynamic.TCPService {
 	servers := make([]dynamic.TCPServer, len(tt.Destination.Pods))
 
+	containMissingPod := false
+
 	for i, podKey := range tt.Destination.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:713]: See my previous comment.

File: pkg/provider/provider.go [hunk 6]
```diff
@@ -730,18 +775,27 @@ func (p *Provider) buildTCPServiceFromTrafficTarget(t *topology.Topology, tt *to
 func (p *Provider) buildWhitelistMiddlewareFromTrafficTargetDirect(t *topology.Topology, tt *topology.ServiceTrafficTarget) *dynamic.Middleware {
 	var IPs []string
 
+	containMissingPod := false
+
 	for _, source := range tt.Sources {
 		for _, podKey := range source.Pods {
 			pod, ok := t.Pods[podKey]
 			if !ok {
-				p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:737]: See my previous comment.

File: pkg/provider/provider.go [hunk 7]
```diff
@@ -755,16 +809,25 @@ func (p *Provider) buildWhitelistMiddlewareFromTrafficTargetDirect(t *topology.T
 func (p *Provider) buildWhitelistMiddlewareFromTrafficSplitDirect(t *topology.Topology, ts *topology.TrafficSplit) *dynamic.Middleware {
 	var IPs []string
 
+	containMissingPod := false
+
 	for _, podKey := range ts.Incoming {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
```
Comment (kevinpollet) [inline pkg/provider/provider.go:761]: See my previous comment.

File: pkg/provider/provider.go (hunk 1 above)
Comment (walbertus) [reply pkg/provider/provider.go:617]: Hi, thanks for your review.

Updated this PR to reduce logging logic.

**Comment by walbertus** [comment]
Need to update missing link in documentation because it's blocking the pipeline.

File: docs/content/install.md [hunk 8]
```diff
@@ -45,7 +45,7 @@ helm install maesh maesh/maesh --set kubedns=true
 ```
 
 With this parameter Maesh will install a CoreDNS as a daemonset.
-KubeDNS will be patched with [stubDomains](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#example-stub-domain).
+KubeDNS will be patched with [stubDomains](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#example).
```
Comment (kevinpollet) [inline docs/content/install.md:48]:
```suggestion
KubeDNS will be patched with [stubDomains](https://v1-17.docs.kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#example-stub-domain).
```

File: pkg/provider/provider.go [hunk 9]
```diff
@@ -639,7 +640,8 @@ func (p *Provider) buildHTTPServiceFromTrafficTarget(t *topology.Topology, tt *t
 	for i, podKey := range tt.Destination.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
+			p.logger.Errorf("Unable to find Pod %q for HTTP service from Service Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```
Comment (kevinpollet) [inline pkg/provider/provider.go:643]:
```suggestion
			p.logger.Errorf("Unable to find Pod %q for HTTP service from Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```

File: pkg/provider/provider.go [hunk 10]
```diff
@@ -710,7 +714,8 @@ func (p *Provider) buildTCPServiceFromTrafficTarget(t *topology.Topology, tt *to
 	for i, podKey := range tt.Destination.Pods {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
+			p.logger.Errorf("Unable to find Pod %q for TCP service from Service Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```
Comment (kevinpollet) [inline pkg/provider/provider.go:717]:
```suggestion
			p.logger.Errorf("Unable to find Pod %q for TCP service from Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```

File: pkg/provider/provider.go [hunk 11]
```diff
@@ -734,7 +739,8 @@ func (p *Provider) buildWhitelistMiddlewareFromTrafficTargetDirect(t *topology.T
 		for _, podKey := range source.Pods {
 			pod, ok := t.Pods[podKey]
 			if !ok {
-				p.logger.Errorf("Unable to find Pod %q", podKey)
+				p.logger.Errorf("Unable to find Pod %q for Middleware from Service Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```
Comment (kevinpollet) [inline pkg/provider/provider.go:742]:
```suggestion
				p.logger.Errorf("Unable to find Pod %q for Middleware from Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```

File: pkg/provider/provider.go (hunk 9 above)
Comment (notsureifkevin) [reply pkg/provider/provider.go]: would `from target:` be unambiguous enough for the purpose of keeping the log volume to a minimum yet provide sufficient context?

File: pkg/provider/provider.go [hunk 12]
```diff
@@ -734,7 +739,8 @@ func (p *Provider) buildWhitelistMiddlewareFromTrafficTargetDirect(t *topology.T
 		for _, podKey := range source.Pods {
 			pod, ok := t.Pods[podKey]
 			if !ok {
-				p.logger.Errorf("Unable to find Pod %q", podKey)
+				p.logger.Errorf("Unable to find Pod %q for Middleware from Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```
Comment (kevinpollet) [inline pkg/provider/provider.go:742]: Sorry, I haven't seen this in my previous review.

```suggestion
				p.logger.Errorf("Unable to find Pod %q for WhitelistMiddleware from Traffic Target %s@%s", podKey, topology.Key{Name: tt.Name, Namespace: tt.Namespace})
```

File: pkg/provider/provider.go [hunk 13]
```diff
@@ -758,7 +764,8 @@ func (p *Provider) buildWhitelistMiddlewareFromTrafficSplitDirect(t *topology.To
 	for _, podKey := range ts.Incoming {
 		pod, ok := t.Pods[podKey]
 		if !ok {
-			p.logger.Errorf("Unable to find Pod %q", podKey)
+			p.logger.Errorf("Unable to find Pod %q for Middleware from Traffic Split %s@%s", podKey, topology.Key{Name: ts.Name, Namespace: ts.Namespace})
```
Comment (kevinpollet) [inline pkg/provider/provider.go:767]:
```suggestion
			p.logger.Errorf("Unable to find Pod %q for WhitelistMiddleware from Traffic Split %s@%s", podKey, topology.Key{Name: ts.Name, Namespace: ts.Namespace})
```

**Pull request merged**